dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers/map-to-fee-record-payment-groups.test.ts

```diff
@@ -0,0 +1,275 @@
+
+      // Act
+      const result = await mapToFeeRecordPaymentGroups(groups);
+
+      // Assert
+      expect(result[0].paymentsReceived).toBeNull();
+    });
+
+    it('sets the totalPaymentsReceived to null', async () => {
+      // Arrange
+      const groups = [createFeeRecordEntityPaymentGroupForSingleFeeRecord(1, 'TO_DO')];
+
+      // Act
+      const result = await mapToFeeRecordPaymentGroups(groups);
+
+      // Assert
+      expect(result[0].totalPaymentsReceived).toBeNull();
+    });
+  });
+
+  describe('when a group has a multiple fee records and payments', () => {
+    it('returns only one fee record payment group', async () => {
+      // Arrange
+      const group: FeeRecordPaymentEntityGroup = {
+        feeRecords: [FeeRecordEntityMockBuilder.forReport(utilisationReport()).build(), FeeRecordEntityMockBuilder.forReport(utilisationReport()).build()],
+        payments: [PaymentEntityMockBuilder.forCurrency('GBP').withAmount(200).build(), PaymentEntityMockBuilder.forCurrency('GBP').withAmount(300).build()],
+      };
+
+      // Act
+      const result = await mapToFeeRecordPaymentGroups([group]);
+
+      // Assert
+      expect(result).toHaveLength(1);
+    });
+
+    it('returns the group with as many fee records as there are fee records in the supplied group', async () => {
+      // Arrange
+      const feeRecordOne = FeeRecordEntityMockBuilder.forReport(utilisationReport()).build();
+      const feeRecordTwo = FeeRecordEntityMockBuilder.forReport(utilisationReport()).build();
+      const group: FeeRecordPaymentEntityGroup = {
+        feeRecords: [feeRecordOne, feeRecordTwo],
+        payments: [PaymentEntityMockBuilder.forCurrency('GBP').build()],
+      };
+
+      // Act
+      const result = await mapToFeeRecordPaymentGroups([group]);
+
+      // Assert
+      expect(result[0].feeRecords).toHaveLength(2);
+    });
+
+    it('sets the totalReportedPayments to the total of the fee record reported payments', async () => {
+      // Arrange
+      const feeRecordOne = FeeRecordEntityMockBuilder.forReport(utilisationReport())
+        .withFeesPaidToUkefForThePeriod(100)
+        .withFeesPaidToUkefForThePeriodCurrency('GBP')
+        .withPaymentCurrency('GBP')
+        .build();
+      const feeRecordTwo = FeeRecordEntityMockBuilder.forReport(utilisationReport())
+        .withFeesPaidToUkefForThePeriod(400)
+        .withFeesPaidToUkefForThePeriodCurrency('JPY')
+        .withPaymentCurrency('GBP')
+        .withPaymentExchangeRate(2)
+        .build();
+      const group: FeeRecordPaymentEntityGroup = {
+        feeRecords: [feeRecordOne, feeRecordTwo],
+        payments: [PaymentEntityMockBuilder.forCurrency('GBP').withAmount(5000000).build()],
+      };
+
+      // Act
+      const result = await mapToFeeRecordPaymentGroups([group]);
+
+      // Assert
+      // 100 + 400 / 2 = 100 + 200 = 300
+      const expectedTotalReportedPaymentAmount = 300;
```

Comment:
Could we use `const` for such values/calculations? Hard to understand this without going through the "mock builder" above. I'm thinking something like this.

```ts
const mockRecordValues = {
  one: {
    period: 100,
    exchangeRate: 1,
  },
  two: {
    period: 200,
    exchangeRate: 2,
  },
};

const feeRecordTwo = FeeRecordEntityMockBuilder.forReport(utilisationReport())
  .withFeesPaidToUkefForThePeriod(mockRecordValues.one.period)
  .withFeesPaidToUkefForThePeriodCurrency('JPY')
  .withPaymentCurrency('GBP')
  .withPaymentExchangeRate(mockRecordValues.one.exchangeRate)
  .build();

const result = await mapToFeeRecordPaymentGroups([group]);

const expetedAmount = mockRecordValues.one.period + mockRecordValues.two.exchangeRate;
```

Appreciate it's not representative of the actual test/values etc

Reply:
I have done this, I'm not sure I personally prefer it and it adds quite a few LOC but can try and do this a bit more going forwards if others find it useful.

Reply:
Thanks @RLavender98 seems a lot clearer to me, especially without the knowledge of the feature/calculations 👍 

Can we add a test for the second fee record? It's only testing the first record at the moment. Then we could remove the `expectedTotalReportedPaymentAmount` so the new values object is purely values

Reply:
`mapToFeeRecordPaymentGroups` returns one object per group, these fee records are grouped and we are passing in just the one group so it returns just one object (but in an array because if you pass it multiple it returns multiple - covered by other tests). Here the total very much depends on both fee records as the total for a group is the sum of the reported payments for the fee records in the group.

To make this more clear have changed the name of the test to use 'sum' and have added an assertion that just one object is returned.

Reply:
Thanks @RLavender98 sorry, can we just extract `expectedTotalReportedPaymentAmount` outside of `testValues` please? As it's an expectation not a mock value

I think the overall tests for this file could be better - for example i'd expect the `calculateTotalCurrencyAndAmount` helper to be called which seems would simplify some things.

However such testing patterns seem to be in a lot of places so not a blocker for this PR.